### PR TITLE
Uniqueness check with different duplicate check strategies.

### PR DESCRIPTION
**Pull request by MishaDemianenko**
Introduce duplicate checking strategies:
use simple map strategy for cases when expected number of elements is
small enough, use dynamic bucket based array strategy in case if expected
number of elements is quite high.
Proposed changes speedups things significantly and for quite big numbers of
duplicates allow create unique indexes in seconds instead of minutes.

### CODE REVIEW

**Comment by MishaDemianenko** [comment]
<b>N.B.: should be null merged to 3.3 because of values refactoring, some changes will be required.</b>

File: community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/index/IndexEntryConflictException.java [hunk 1]
```diff
@@ -38,7 +38,7 @@
     private final long addedNodeId;
     private final long existingNodeId;
 
-    public IndexEntryConflictException( long existingNodeId, long addedNodeId, Object propertyValue )
+    public IndexEntryConflictException( long existingNodeId, long addedNodeId, Object... propertyValue )
```
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/index/IndexEntryConflictException.java]: Could use `propertyValues` to reflect var args in parameter name.

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+                {
+                    return true;
+                }
+                if ( o == null || getClass() != o.getClass() )
+                {
+                    return false;
+                }
+
+                PropertyValues that = (PropertyValues) o;
+                return propertyValuesEqual( properties, that.values );
```
Comment (burqen) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java:136]: `equals` and `hashCode` does not align. Both property key and value is used to calculate `hashCode` but equals only check property value, which means two `PropertyValues` object can be equal but have different hash code. For the use case that we have this is fine, but confusing. 

Suggestion: Change the hash code function to only check values array instead to make it less confusing.

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+     */
+    static class BucketsDuplicateCheckStrategy extends DuplicateCheckStrategy
+    {
+        private static final int BASE_ENTRY_SIZE = 1000;
+        private static final int DEFAULT_BUCKETS = 10;
+        static final int BUCKET_STRATEGY_ENTRIES_THRESHOLD = BASE_ENTRY_SIZE * DEFAULT_BUCKETS;
+
+        private static final int MAX_NUMBER_OF_BUCKETS = 100;
+        private final int numberOfBuckets;
+        private EntrySet[] actualValues;
```
Comment (burqen) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java]: I was a bit confused about what was the "buckets". Can we improve naming to make this more clear?

Something like (but probably not):
`actualValues` -> `buckets`
`EntrySet` -> `BucketNode`
`entrySetSize` -> `bucketNodeSize`

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            EntrySet current = actualValues[bucket];
+            if ( current == null )
+            {
+                current = new EntrySet( entrySetSize );
+                actualValues[bucket] = current;
+            }
+            return current;
+        }
+
+        private static class EntrySet
```
Comment (burqen) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/DuplicateCheckStrategy.java]: A small javadoc about this guy being a linked list where each entry contain `nodeId`:`propertyValue` pairs would be nice.

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/PartitionedUniquenessVerifier.java [hunk 5]
```diff
@@ -133,18 +131,55 @@ private Terms termsForField( String fieldName ) throws IOException
...
-             * Here {@link DuplicateCheckingCollector#reset()} is deliberately not called to preserve accumulated
-             * state (knowledge about duplicates) across all {@link IndexSearcher#search(Query, Collector)} calls.
-             */
-            SimpleCollector collector = DuplicateCheckingCollector.forProperties( accessor, propertyKeyIds );
             for ( PartitionSearcher searcher : searchers )
             {
+                /*
+                 * Here {@link DuplicateCheckingCollector#init}} is deliberately not called to preserve accumulated
+                 * state (knowledge about duplicates) across all {@link IndexSearcher#search(Query, Collector)} calls.
+                 */
```
Comment (burqen) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/PartitionedUniquenessVerifier.java:182]: Nice clarification!

**Review by burqen — CHANGES_REQUESTED** [review]
The map approach has O(1) in time complexity for `checkForDuplicate`.
Bucket approach has O(n).
This makes a little bit worried. Do we have any benchmarks that verifies that this is not a significant regression?

**Comment by MishaDemianenko** [comment]
@burqen updated

**Review by burqen — APPROVED** [review]

**Pull request merged**